apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -584,19 +584,19 @@ wait_for_feature_before_auth({xmlstreamelement, El}, StateData) ->
                                           });
         {?NS_COMPRESS, <<"compress">>} when Zlib == true,
                                             ((SockMod == gen_tcp) or
-                                             (SockMod == ejabberd_tls)) ->
+                                             (SockMod == fast_tls)) ->
           check_compression_auth(El, wait_for_feature_before_auth, StateData);
         _ ->
-            if
-                TLSRequired and not TLSEnabled ->
-                    Lang = StateData#state.lang,
-                    send_element(StateData, ?POLICY_VIOLATION_ERR(
-                                               Lang, <<"Use of STARTTLS required">>)),
-                    send_trailer(StateData),
-                    {stop, normal, StateData};
-                true ->
-                    process_unauthenticated_stanza(StateData, El),
-                    fsm_next_state(wait_for_feature_before_auth, StateData)
+            case (TLSRequired and not TLSEnabled) of
```

Comment:
Since we are touching this code, can we move it to a function `terminate_when_tls_required_but_not_enabled`?
I know the name is quite long, but it says what it's doing.